### DISCUSSION TITLE
How to conditionally load sbt projects?

### DISCUSSION THREAD

**Discussion by gontard** [sbt/sbt#7214] [Q&A] [answered]
Given this simple multi-module project:

```
lazy val module1 = (project in file("module1")).settings(name := "module1")
lazy val module2 = (project in file("module2")).settings(name := "module2")
lazy val module3 = (project in file("module3")).settings(name := "module3")

lazy val root = (project in file("."))
  .dependsOn(module1, module2, module3)
  .aggregate(module1, module2, module3)
  .settings(
    name := "sbt-conditional-project-loading"
  )
```

For performance reasons, I want to be able to load only a subset of the modules. For example, using an environment variable:

```
$ sbt projects
[info] 	 * root
$ MODULES=module1,module2 sbt projects
[info] 	   module1
[info] 	   module2
[info] 	 * root
```

The code is available here: https://github.com/gontard/sbt-conditional-project-loading

Is there a way to achieve that?

Any help is greatly appreciated

**Comment by SethTisue** [comment, marked as answer]
maybe if you create them this way? https://stackoverflow.com/questions/66367587/dynamically-create-subprojects-in-scala-sbt

I'm just guessing. I really don't think there's any way if the projects are declared statically, but perhaps if you create them dynamically you could do it. you might see if there are other Stack Overflow questions about dynamic project creation

lord knows what problems making all the projects dynamically might cause with e.g. IDE integration? :shrug:

**Reply by gontard** [reply]
Thanks for your help! Indeed, it works using a `CompositeProject` (even if I don't understand why):

```scala
val moduleNames: List[String] = sys.env.getOrElse("MODULES", "").split(",").filter(_.nonEmpty).toList
val subprojects = new CompositeProject {
  override def componentProjects: Seq[Project] = {
    moduleNames.map(file).map { p =>
      Project(p.getName, p).settings(name := p.getName)
    }
  }
}

val projects: List[Project] = subprojects.componentProjects.toList
val modules: List[ClasspathDep[ProjectReference]] =
  projects.map(classpathDependency(_)(Project.projectToRef))


lazy val root = (project in file("."))
  .dependsOn(modules: _*)
  .aggregate(modules.map(_.project): _*)
  .settings(
    name := "sbt-conditional-project-loading"
  )
```

> lord knows what problems making all the projects dynamically might cause with e.g. IDE integration? 🤷

It works correctly in my test project but yes, I need to test on a real-world project
